### PR TITLE
Improve Symfony\Component\HttpFoundation\Request

### PR DESCRIPTION
**Pull request by yyaremenko**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #...   
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

- use of constants for HTTP verbs names instead of strings
- use of constants for HTTP / HTTPS ports
- PHPDoc update for some properties
- some constructors incoming parameters are split into multiple lines
- method getLanguages(): made readable, nesting level reduced
- getLanguage() test improved to use dataProvider

### CODE REVIEW

**Comment by yceruto** [comment]
Hi @yyaremenko thanks for your PR.

Symfony defines some coding standards that all contributions must follow. e.g:

> some constructors incoming parameters are split into multiple lines

It should be reverted according to this statement:

> Declare all the arguments on the same line as the method/function name, no matter how many arguments there are;

Please take a look to see if anything else is missing:
https://symfony.com/doc/current/contributing/code/standards.html

**Comment by yceruto** [comment]
By the way, code refactoring should be targeted for `master` branch (as far as I remember).

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 1]
```diff
@@ -58,6 +58,14 @@ class Request
     const METHOD_TRACE = 'TRACE';
     const METHOD_CONNECT = 'CONNECT';
 
+    const HEADER_REQUEST_METHOD = 'REQUEST_METHOD';
+    const HEADER_METHOD_OVERRIDE = 'X-HTTP-METHOD-OVERRIDE';
+    const HEADER_CONTENT_TYPE = 'CONTENT_TYPE';
+    const HEADER_REQUEST_URI = 'REQUEST_URI';
+
+    const PORT_HTTPS = 443;
+    const PORT_HTTP = 80;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: all these constants should be removed: adding them would make them part of the public API, which means they would be covered by the BC promise. That's not desired.

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 2]
```diff
@@ -250,8 +258,15 @@ class Request
      * @param array                $server     The SERVER parameters
      * @param string|resource|null $content    The raw body data
      */
-    public function __construct(array $query = [], array $request = [], array $attributes = [], array $cookies = [], array $files = [], array $server = [], $content = null)
-    {
+    public function __construct(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: all such changes must be reverted: our CS is preferring long lines over split ones.

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 3]
```diff
@@ -1746,30 +1807,37 @@ public function getLanguages()
             return $this->languages;
         }
 
-        $languages = AcceptHeader::fromString($this->headers->get('Accept-Language'))->all();
+        $languageHeaderValue = $this->headers->get('Accept-Language');
+        $languages = array_keys(AcceptHeader::fromString($languageHeaderValue)->all());
+
+        $delimiter = '-';
         $this->languages = [];
-        foreach ($languages as $lang => $acceptHeaderItem) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: Please revert the change in this file: the code works fine, changing them for CS reasons can lead to introducing bugs. We already merged similar changes and broke things by mistake.

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 4]
```diff
@@ -2106,10 +2174,11 @@ private function getTrustedValues($type, $ip = null)
     {
         $clientValues = [];
         $forwardedValues = [];
+        $defaultIp = '0.0.0.0';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Request.php]: should be reverted: this added variable provides nothing in practice

File: src/Symfony/Component/HttpFoundation/Tests/RequestTest.php [hunk 5]
```diff
@@ -1543,31 +1543,35 @@ public function testGetAcceptableContentTypes()
         $this->assertEquals(['application/vnd.wap.wmlscriptc', 'text/vnd.wap.wml', 'application/vnd.wap.xhtml+xml', 'application/xhtml+xml', 'text/html', 'multipart/mixed', '*/*'], $request->getAcceptableContentTypes());
     }
 
-    public function testGetLanguages()
+    /**
+     * @return array
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Tests/RequestTest.php]: we don't document tests

File: src/Symfony/Component/HttpFoundation/Tests/RequestTest.php [hunk 6]
```diff
@@ -1543,31 +1543,35 @@ public function testGetAcceptableContentTypes()
...
+     */
+    public function languagesTestDataProvider()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Tests/RequestTest.php]: please move the provider after the consumer

File: src/Symfony/Component/HttpFoundation/Tests/RequestTest.php [hunk 7]
```diff
@@ -1543,31 +1543,35 @@ public function testGetAcceptableContentTypes()
...
+            ['zh, i-cherokee; q=0.6', ['zh', 'cherokee']],
+            ['i-imaginary-lang, a-never-existing-lang; q=0.6', ['imaginary', 'a_NEVER_EXISTING_LANG']], // Imaginary languages, to validate specific cases
+        ];
+    }
 
+    /**
+     * @dataProvider languagesTestDataProvider
+     *
+     * @param string $languageHeaderValue
+     * @param array  $expectedResult
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Tests/RequestTest.php]: should be removed (`array` can be added to the real signature)

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 4 above)
Comment (yyaremenko) [reply src/Symfony/Component/HttpFoundation/Request.php]: yes, it does provide. a code must never use a hardcoded repeating string - at least such string must be placed in a private constant; but, as your code does not use private constants, as far as I can see, I decided to assign it to a variable inside the method.

File: src/Symfony/Component/HttpFoundation/Request.php (hunk 3 above)
Comment (yyaremenko) [reply src/Symfony/Component/HttpFoundation/Request.php]: no, I won't revert. 

if you have the code, covered with tests, that you are 'afraid to touch, because something can go wrong' (no matter the reason of changes, whether it is code style or functionality) - you have either a bad code or lack of tests. 

if you looked at that code part, you would notice the variable $acceptHeaderItem was never used. but no, let's leave it as is. good approach! but not good enough for me, sadly.

**Comment by yyaremenko** [comment]
so, in conclusion, I could only repeat my comment for another PR I made - I am really, really shocked with the code quality of Symfony. and the fact a lot of contributors fiercely resisting any logical, sane code improvements.

so be it.
